core-services/workflow/src/main/java/org/silverpeas/core/workflow/engine/TaskSavedRequest.java

```diff
@@ -0,0 +1,115 @@
+    TaskSavedRequest request = ServiceProvider.getService(TaskSavedRequest.class);
+    request.event = event;
+    return request;
+  }
+
+  @Override
+  public void process(final Object context) throws InterruptedException {
+
+    // Get the process instance
+    UpdatableProcessInstance instance = (UpdatableProcessInstance) event.getProcessInstance();
+    String id = instance.getInstanceId();
+    ProcessInstanceManager instanceManager = WorkflowHub.getProcessInstanceManager();
+    Mutable<UpdatableHistoryStep> step = Mutable.of(null);
+
+    Transaction.performInOne(()-> {
+      UpdatableHistoryStep newStep = null;
+      UpdatableProcessInstance processInstance = getProcessInstanceRepository().getById(id);
+
+      if (!event.isFirstTimeSaved()) {
+        // reload historystep that has been already created
+        newStep = (UpdatableHistoryStep) processInstance.getSavedStep(event.getUser().getUserId());
+      } else {
+        // first create the history newStep
+        newStep = (UpdatableHistoryStep) instanceManager.createHistoryStep();
+        newStep.setUserId(event.getUser().getUserId());
+        newStep.setAction(event.getActionName());
+        newStep.setActionDate(event.getActionDate());
+        newStep.setUserRoleName(event.getUserRoleName());
+        if (event.getResolvedState() != null) {
+          newStep.setResolvedState(event.getResolvedState().getName());
+        }
+        // To be processed
+        newStep.setActionStatus(0);
+        newStep.setProcessInstance(processInstance);
+
+        // add the new newStep to the processInstance
+        getHistoryStepRepository().save((HistoryStepImpl) newStep);
+      }
+      step.set(newStep);
+      return newStep;
+    });
+
+    Transaction.performInOne(()-> {
+      UpdatableProcessInstance processInstance = getProcessInstanceRepository().getById(id);
+
+      // Do workflow stuff
+      try {
+        processEvent(processInstance, step.get().getId(), event.getResolvedState());
+        getProcessInstanceRepository().save((ProcessInstanceImpl) processInstance);
+      } catch (WorkflowException we) {
+        saveError(processInstance, event, step.get(), we);
+
+        throw new WorkflowException("WorkflowEngineThread.process",
+            "workflowEngine.EX_ERR_PROCESS_ADD_TASKSAVED_REQUEST", we);
+      }
+
+      return null;
+    });
+  }
+
+  /**
+   * Method declaration
+   */
+  private boolean processEvent(UpdatableProcessInstance instance, String stepId, State state)
+      throws WorkflowException {
+
+    UpdatableHistoryStep step = (UpdatableHistoryStep) instance.getHistoryStep(stepId);
+    // only to set the current step of instance to that step
+    instance.updateHistoryStep(step);
+
+    // Saving data of step and process instance
+    if (event.getDataRecord() != null) {
+      instance.saveActionRecord(step, event.getDataRecord());
+    }
+
+    // Add current user as working user
+    state = (state != null) ? state : new StateImpl("");
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Introduce a new variable instead of reusing the parameter "state". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1226)